components/controls/Tooltip.tsx

```diff
@@ -262,35 +305,47 @@ export class TooltipInner extends React.Component<TooltipProps, State> {
         })}
         {this.isVisible() && (
           <TooltipPortal>
-            <ScreePositionFixer ready={isMeasured(this.state)}>
-              {({ leftFix = 0, topFix = 0 }) => (
-                <div
-                  className={`${classNameSpace} ${this.getPlacement()}`}
-                  onMouseEnter={this.handleOverlayMouseEnter}
-                  onMouseLeave={this.handleOverlayMouseLeave}
-                  ref={this.tooltipNodeRef}
-                  style={
-                    isMeasured(this.state)
-                      ? {
-                          left: this.state.left + leftFix,
-                          top: this.state.top + topFix,
-                          width: this.state.width,
-                          height: this.state.height
-                        }
-                      : undefined
-                  }>
-                  <div className={`${classNameSpace}-inner`}>{this.props.overlay}</div>
-                  <div
-                    className={`${classNameSpace}-arrow`}
-                    style={
-                      isMeasured(this.state)
-                        ? { marginLeft: -leftFix, marginTop: -topFix }
-                        : undefined
+            <ScreenPositionFixer ready={isMeasured(this.state)}>
+              {({ leftFix = 0, topFix = 0 }) => {
+                let placement = this.getPlacement();
+                let style = isMeasured(this.state)
+                  ? {
+                      left: this.state.left + leftFix,
+                      top: this.state.top + topFix,
+                      width: this.state.width,
+                      height: this.state.height
                     }
-                  />
-                </div>
-              )}
-            </ScreePositionFixer>
+                  : undefined;
+
+                if (style && !this.props.forcePlacement) {
+                  const adjustment = this.adjustTooltipPosition(placement, style, {
+                    leftFix,
+                    topFix
+                  });
+                  style = adjustment.style;
+                  placement = adjustment.placement;
+                }
```

Comment:
Also, as discussed, extracting the logic from the render function might be great for readability.